src/loader/riscv/pe/mod.rs

```diff
@@ -0,0 +1,253 @@
+    ///
+    /// # Arguments
+    ///
+    /// * `guest_mem` - The guest memory where the kernel image is loaded.
+    /// * `kernel_offset` - 2MiB-aligned (for riscv64) base address in guest memory at which to load the kernel.
+    /// * `kernel_image` - Input Image format kernel image.
+    /// * `highmem_start_address` - ignored on RISC-V.
+    ///
+    /// # Returns
+    /// * KernelLoaderResult
+    fn load<F, M: GuestMemory>(
+        guest_mem: &M,
+        kernel_offset: Option<GuestAddress>,
+        kernel_image: &mut F,
+        _highmem_start_address: Option<GuestAddress>,
+    ) -> Result<KernelLoaderResult>
+    where
+        F: Read + Seek,
+    {
+        let kernel_size = kernel_image
+            .seek(SeekFrom::End(0))
+            .map_err(|_| Error::SeekImageEnd)? as usize;
+        let mut riscv_header: riscv_image_header = Default::default();
+        kernel_image.rewind().map_err(|_| Error::SeekImageHeader)?;
+
+        riscv_header
+            .as_bytes()
+            .read_from(0, kernel_image, mem::size_of::<riscv_image_header>())
+            .map_err(|_| Error::ReadImageHeader)?;
+
+        // Skip testing `magic` field as it's deprecated as of RISC-V boot image header version 0.2
+        // if u64::from_le(riscv_header.magic) != 0x5643534952 {
+        //     return Err(Error::InvalidImageMagicNumber.into());
+        // }
+
+        if u32::from_le(riscv_header.magic2) != 0x05435352 {
+            return Err(Error::InvalidImageMagicNumber.into());
+        }
+
+        let text_offset = u64::from_le(riscv_header.text_offset);
+
+        // Validate that kernel_offset is 2MiB aligned (for riscv64)
+        #[cfg(target_arch = "riscv64")]
+        if let Some(kernel_offset) = kernel_offset {
+            if kernel_offset.raw_value() % 0x0020_0000 != 0 {
+                return Err(Error::InvalidBaseAddrAlignment.into());
+            }
+        }
+
+        let mem_offset = kernel_offset
+            .unwrap_or(GuestAddress(0))
+            .checked_add(text_offset)
+            .ok_or(Error::InvalidImage)?;
+
+        let mut loader_result = KernelLoaderResult {
+            kernel_load: mem_offset,
+            ..Default::default()
+        };
+
+        kernel_image.rewind().map_err(|_| Error::SeekImageHeader)?;
+        guest_mem
+            .read_exact_from(mem_offset, kernel_image, kernel_size)
+            .map_err(|_| Error::ReadKernelImage)?;
+
+        loader_result.kernel_end = mem_offset
+            .raw_value()
+            .checked_add(kernel_size as GuestUsize)
+            .ok_or(KernelLoaderError::MemoryOverflow)?;
+
+        Ok(loader_result)
+    }
+}
+
+/// Writes the device tree to the given memory slice.
+///
+/// # Arguments
+///
+/// * `guest_mem` - A u8 slice that will be partially overwritten by the device tree blob.
+/// * `guest_addr` - The address in `guest_mem` at which to load the device tree blob.
+/// * `dtb_image` - The device tree blob.
+pub fn load_dtb<F, M: GuestMemory>(
```

Comment:
Not really sure if this functionality should be here - it's just using vm-memory to load into the memory. Something any VMM can easily do. What motivated you to add this here?

Reply:
Resolved by removing load_dtb() since it's doable using `vm-memory`